astropy/io/fits/hdu/hdulist.py

```diff
@@ -694,14 +694,13 @@ def writeto(self, fileobj, output_verify='exception', overwrite=False,
         # but only if the file doesn't exist.
         fileobj = _File(fileobj, mode='ostream', overwrite=overwrite)
         hdulist = self.fromfile(fileobj)
+        dirname = os.path.dirname(hdulist._file.name)
 
-        for hdu in self:
-            hdu._prewriteto(checksum=checksum)
-            try:
+        with _free_space_check(self, dirname=dirname):
+            for hdu in self:
+                hdu._prewriteto(checksum=checksum)
                 hdu._writeto(hdulist._file)
-            finally:
```

Comment:
Yes, I think the try/finally in hdulist.close() should be able to deal with all kinds of interrupted write, given that the file was opened with ``with``, or the close() was called manually. 

As for the individual hdu's, the exception handling in ``_free_space_check()`` now calls hdu._close() for all of them.